File: java_test/src/DesignPattern/TemplateMethod/SimpleBarista/Coffee.java
package DesignPattern.TemplateMethod.SimpleBarista;

/**
 * Created by Administrator on 2018/4/28.
 */
public class Coffee {
    //准备食谱
    void prepareRecipe() {
        boilWater();
        brewCoffeeGrinds();
        pourInCup();
        addSugarAndMilk();
    }

    //烧水
    public void boilWater() {
        System.out.println("Boiling water");
    }

    //冲咖啡粉
    public void brewCoffeeGrinds() {
        System.out.println("Dripping Coffee through filter");
    }

    //倒进杯子
    public void pourInCup() {
        System.out.println("Pouring into cup");
    }

    //添加糖和牛奶
    public void addSugarAndMilk() {
        System.out.println("Adding Sugar and Milk");
    }
}

File: test_go/src/pattern/09_template_method/eg2/coffee.go
package eg2

import "fmt"

type Coffee struct {
	baseCaffeineBeverage
}

func NewCoffee() *Coffee {
	this := new(Coffee)
	this.CaffeineBeverage = this
	return this
}

func (this *Coffee) Brew() {
	fmt.Println("Dripping Coffee through filter")
}

func (this *Coffee) AddCondiments() {
	fmt.Println("Adding Sugar and Milk")
}
